_posts/2024-08-10-integration-for-kw-ssh.md

```diff
@@ -0,0 +1,112 @@
+processes work correctly.
+
+![Picture](/images/containerfile_ssh.png)
+
+# Challenges with Nested Containers
+
+When creating containers within containers, executing commands directly from
+the host machine to the nested container becomes a challenge. To address this
+complexity, I developed the `container_exec_in_nested_container()` function,
+which facilitates command execution within this nested environment.
+
+![Picture](/images/container_nested.png)
+
+This function facilitates the execution of commands in a nested container
+within another container, which is common in kw-ssh integration tests. It uses
+the `container_exec()` function, which executes commands directly in a
+container. One of the challenges when passing commands to containers is
+handling special characters, such as single quotes, which can cause issues
+during execution. To address this, I used the `str_escape_single_quotes()`
+function, which correctly escapes these characters, ensuring that commands are
+executed reliably.
+
+# Managing Commands in Nested Containers
+
+The `str_escape_single_quotes()` function uses the **sed** command to add
+backslashes `(\)` before any single quotes found in the string, allowing
+commands containing single quotes to be interpreted correctly by the shell:
+
+![Picture](/images/str_escape_single_quotes.png)
+
+Additionally, in the `container_exec_in_nested_container()` function, I use the
+special `$''` format for strings, known as "ANSI C quoting". This format allows
+escape sequences such as `\'` (escaped single quotes) to be processed correctly
+by the shell. The use of `$''` is essential here to ensure that commands are
+interpreted correctly, even when they contain characters that would otherwise
+need to be escaped. This prevents errors when running tests.
+
+Here's an example of how this approach is implemented:
+
+
+```bash
+cmd+=" ${inner_container_name} /bin/bash -c $'${inner_container_command}'"
+```
+By using `$'`, the string passed to the container can contain special
+characters without causing problems at runtime. This is especially important
+when working with nested containers, where proper string handling is critical
+to the success of integration tests.
+
+
+# Integration Test Example with kw-ssh
+
+![Picture](/images/kw_ssh_test_example.png)
+
+This test example verifies the SSH connection of `kw-ssh` using the remote
+connections configuration file. Typically, this file can be found at
+*~/.config/kw/remote.config*. This test illustrates how the
+`container_exec_in_nested_container()` function is used to manage command
+execution in nested containers and how the test is conducted across different
+Linux distributions.
+
+# Conclusion
+
+Integration tests for kw-ssh ensure that the feature works correctly across
+different Linux distributions. Using containers to isolate test environments
+and the SSH server allows for precise and consistent validation. The functions
```

Comment:
"With the isolation of test environments from containers in conjunction with an SSH server, we achieve precise and consistent validation."